Fixes #2548 : Makes InOrder able to verify static methods

mirkoalicastro opened the pull request:
<!-- Hey,
Thanks for the contribution, this is awesome.
As you may have read, project members have somehow an opinionated view on what and how should be
Mockito, e.g. we don't want mockito to be a feature bloat.
There may be a thorough review, with feedback -> code change loop.
-->
<!--
If you have a suggestion for this template you can fix it in the .github/PULL_REQUEST_TEMPLATE.md file
-->

This PR makes it possible to verify static methods calls in order. Fixes: #2548 

Example of usage:

```
try (MockedStatic<Foo> mocked = mockStatic(Foo.class)) {
  InOrder inOrder = inOrder(Foo.class);
  mocked.when(Foo::firstMethod).thenReturn("first");
  assertEquals("first", Foo.firstMethod());

  inOrder.verify(mocked, Foo::firstMethod);
}
```
You can also pass a specific `VerificationMode` as third argument to the `InOrder#verify` method.

## Checklist

 - [x] Read the [contributing guide](https://github.com/mockito/mockito/blob/main/.github/CONTRIBUTING.md)
 - [x] PR should be motivated, i.e. what does it fix, why, and if relevant how
 - [x] If possible / relevant include an example in the description, that could help all readers
       including project members to get a better picture of the change
 - [x] Avoid other runtime dependencies
 - [x] Meaningful commit history ; intention is important please rebase your commit history so that each
       commit is meaningful and help the people that will explore a change in 2 years
 - [x] The pull request follows coding style
 - [x] Mention `Fixes #<issue number>` in the description _if relevant_
 - [x] At least one commit should mention `Fixes #<issue number>` _if relevant_

TimvdLippe commented on subprojects/inline/src/test/java/org/mockitoinline/InOrderVerificationTest.java:
```diff
@@ -0,0 +1,191 @@
+        });
+    }
+
+    @Test(expected = VerificationInOrderFailure.class)
```
Nit: do not use `@Test(expected)` and instead use `org.assertj.core.api.Assertions.assertThatThrownBy`. We recently migrated away from that in #2510 (I should probably check if we can enable ErrorProne to error on this).

TimvdLippe reviewed (changes requested):
Thanks for the PR! Overall, the implementation looks solid. I only have a couple of nits with regards to the tests.

My only concern is the difference in the class passed into `inOrder` and the class we are actually interacting with. I guess this currently works, as we are passing in a `Class` in the `List<Object>`, but that feels a bit weird to me. Not sure WDYT?

I think to fix that, we would need a similar method to `objectIsMockToBeVerified`, but in this case verify that the context is correct. However, we are only passing in the lambda, which doesn't necessarily match to the `Class` we put in.

Therefore, we maybe have to look into passing the `mockedStatic` in the `inOrder` method instead and omit it as an argument from the `inOrder.verify` call?

mirkoalicastro commented on subprojects/inline/src/test/java/org/mockitoinline/InOrderVerificationTest.java:
```diff
@@ -0,0 +1,191 @@
+import org.mockito.exceptions.verification.VerificationInOrderFailure;
+import org.mockito.verification.VerificationMode;
+
+public class InOrderVerificationTest {
```
Definitely, it looks a great test to add. Done!

mirkoalicastro commented on subprojects/inline/src/test/java/org/mockitoinline/InOrderVerificationTest.java:
```diff
@@ -0,0 +1,191 @@
+        });
+    }
+
+    private void mockedStaticTest(Consumer<MockedStatic<StaticContext>> body) {
```
Sure, I see your point. I have inlined that in the tests.

mirkoalicastro commented:
> Thanks for the PR! Overall, the implementation looks solid. I only have a couple of nits with regards to the tests.
> 
> My only concern is the difference in the class passed into `inOrder` and the class we are actually interacting with. I guess this currently works, as we are passing in a `Class` in the `List<Object>`, but that feels a bit weird to me. Not sure WDYT?
> 
> I think to fix that, we would need a similar method to `objectIsMockToBeVerified`, but in this case verify that the context is correct. However, we are only passing in the lambda, which doesn't necessarily match to the `Class` we put in.
> 
> Therefore, we maybe have to look into passing the `mockedStatic` in the `inOrder` method instead and omit it as an argument from the `inOrder.verify` call?

Thanks for your feedback @TimvdLippe ! I have addressed the comments and pushed the changes related to the tests.

I see your concern, and actually, passing the `mockedStatic` to the `inOrder` method was my initial idea. Here is the reason why I moved away from that.
Let's assume we have two static mocks (two different classes), and we want to verify them in order. Then, we'd pass two `MockedStatic` instances to the `inOrder` method.
When it comes to the verification, we'd have `InOrder#verify(Verification)`, where `Verification` is a lambda (no parameters, void return type) because `MockedStatic#verify` deals with a lambda. Given a lambda, we can't assert if it contains a static method reference, or contains just some statements. That is, we can't retrieve the `Class` instance. Without this, we can't select and use the appropriate `MockedStatic` instance to verify the lambda.
Moreover, the above assumes we have the `Class` instance of the `MockedStatic` somehow available in `InOrderImpl`, but currently, that is not the case. Of course, we could add a method to `MockedStatic` that accepts a `Class` and returns true if the `Class` is the same as what is mocking (package private maybe?). 
I see your concern, but I think we have the same situation in the `MockedStatic#verify` methods, where we could pass things not relevant to the context. But I can't see a different way to solve it otherwise, if possible.
WDYT?

TimvdLippe reviewed (approved):
Awesome, thanks for the explanation (makes sense to me) and the nice PR + tests! Merging 🎉
